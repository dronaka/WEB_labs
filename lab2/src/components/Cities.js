import React from "react";
import { connect } from "react-redux";

import CityForm from "./CityForm"
import City from "./City";
import { addСity, deleteСity, fetchWeatherByCityName } from "../actions";



class Сities extends React.Component {
  render() {
    return (
      <div className="Сities">
        <h1>Сities</h1>
        <CityForm onSubmit={(e) => this.handleCityForm(e)} />
        {this.props.error && <div className="error">Error: {this.props.error}</div>}
        <div className="forecasts">
          {
            [...this.props.cities.entries()].map((entry) => {
              console.log(entry[0])
              return (
                <City
                  key={entry[0]}
                  onFetch={() => this.props.fetchWeatherByCityName(entry[0])}
                  onDelete={() => this.props.deleteСity(entry[0])}
                  forecast={entry[1]} />
              );
            })
          }
        </div>
      </div>
    );
  }

  handleCityForm(e) {
    e.preventDefault();
    const cityName = e.currentTarget.elements.cityName.value;
    this.props.addСity(cityName);
  }
}


function mapStateToProps(state) {
  return {
    cities: state.cities,
    error: state.error
  };
}

function mapDispatchToProps(dispatch) {
  return {
    addСity: (cityName) => {
      dispatch(addСity(cityName));
    },

    deleteСity: (cityName) => {
      dispatch(deleteСity(cityName)); 
    },

    fetchWeatherByCityName: (cityName) => {
      dispatch(fetchWeatherByCityName(cityName));
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Сities);