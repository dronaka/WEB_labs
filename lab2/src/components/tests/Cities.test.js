import React from "react";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import Cities from "../Cities";


describe("<Cities /> render", () => {

  const DEFAULT_FORECAST = {
    cityName: "New York",
    temperature: "-22",
    icon: "icon_code",
    windSpeed: 15,
    description: "good to stay at home",
    pressure: 234,
    humidity: 1,
    coords: {
      lat: 74,
      lon: 89
    }
  };

  const storeCreator = configureStore([thunk]);

  beforeAll(() => {
    global.fetch = jest.fn(() => Promise.resolve({
      ok: true,
      json: () => Promise.resolve("fake success response")}
      ))
  });

  afterAll(() => {
    delete global.fetch;
  });

  test("should has error and empty cities", () => {
    const store = storeCreator({
    cities: {
        error: "test error message",
        cities: new Map()
      }
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Cities />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should has cities with 1 loading element", () => {
    const store = storeCreator({
      cities: {
        cities: new Map([["firstLoadingCity", undefined]])
      }
    });

    const tree = renderer
    .create(
      <Provider store={store}>
        <Cities />
      </Provider>
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should has error and cities with 1 element", () => {
    const store = storeCreator({
      cities: {
        error: "test error message",
        cities: new Map([["firstCity", DEFAULT_FORECAST]])
      }
    });

    const tree = renderer
    .create(
      <Provider store={store}>
        <Cities />
      </Provider>
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should has empty cities", () => {
    const store = storeCreator({
      cities: {
        cities: new Map()
      }
    });

    const tree = renderer
    .create(
      <Provider store={store}>
        <Cities />
      </Provider>
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should has cities with 1 element", () => {
    const store = storeCreator({
    cities: {
        cities: new Map([["firstCity", DEFAULT_FORECAST]])
      }
    });

    const tree = renderer
    .create(
      <Provider store={store}>
        <Cities />
      </Provider>
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should has cities with 2 element", () => {
    const store = storeCreator({
    cities: {
        cities: new Map([
          ["firstCity", DEFAULT_FORECAST],
          ["secondCity", DEFAULT_FORECAST]])
      }
    });

    const tree = renderer
    .create(
      <Provider store={store}>
        <Cities />
      </Provider>
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

});