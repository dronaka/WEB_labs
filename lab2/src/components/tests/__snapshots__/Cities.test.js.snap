// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cities /> render should has cities with 1 element 1`] = `
<div
  className="cities"
>
  <h1>
    Сities
  </h1>
  <form
    class="add-city"
    onSubmit={[Function]}
  >
    <input
      class="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      class="button"
      type="submit"
      value="Add to city"
    />
  </form>
  <div
    className="forecasts"
  >
    <div
      className="city"
    >
      <div
        className="header"
      >
        <div
          className="city-name"
        >
          New York
        </div>
        <div
          className="temperature"
        >
          -22
           ℃
        </div>
        <img
          alt="Weather icon"
          src="https://openweathermap.org/img/wn/icon_code.png"
        />
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Wind
        </div>
        <div
          className="value"
        >
          15 m/s
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Cloudness
        </div>
        <div
          className="value"
        >
          good to stay at home
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Pressure
        </div>
        <div
          className="value"
        >
          234 hPa
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Humidity
        </div>
        <div
          className="value"
        >
          1%
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Coords
        </div>
        <div
          className="value"
        >
          74, 89
        </div>
      </div>
      <button
        className="button"
        onClick={[Function]}
      >
        delete
      </button>
    </div>
  </div>
</div>
`;

exports[`<Cities /> render should has cities with 1 loading element 1`] = `
<div
  className="cities"
>
  <h1>
    Сities
  </h1>
  <form
    class="add-city"
    onSubmit={[Function]}
  >
    <input
      class="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      class="button"
      type="submit"
      value="Add to city"
    />
  </form>
  <div
    className="forecasts"
  >
    <div
      className="loader"
    >
      <p>
        Forecast is loading...
      </p>
    </div>
  </div>
</div>
`;

exports[`<Cities /> render should has cities with 2 element 1`] = `
<div
  className="cities"
>
  <h1>
    Сities
  </h1>
  <form
    class="add-city"
    onSubmit={[Function]}
  >
    <input
      class="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      class="button"
      type="submit"
      value="Add to city"
    />
  </form>
  <div
    className="forecasts"
  >
    <div
      className="city"
    >
      <div
        className="header"
      >
        <div
          className="city-name"
        >
          New York
        </div>
        <div
          className="temperature"
        >
          -22
           ℃
        </div>
        <img
          alt="Weather icon"
          src="https://openweathermap.org/img/wn/icon_code.png"
        />
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Wind
        </div>
        <div
          className="value"
        >
          15 m/s
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Cloudness
        </div>
        <div
          className="value"
        >
          good to stay at home
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Pressure
        </div>
        <div
          className="value"
        >
          234 hPa
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Humidity
        </div>
        <div
          className="value"
        >
          1%
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Coords
        </div>
        <div
          className="value"
        >
          74, 89
        </div>
      </div>
      <button
        className="button"
        onClick={[Function]}
      >
        delete
      </button>
    </div>
    <div
      className="city"
    >
      <div
        className="header"
      >
        <div
          className="city-name"
        >
          New York
        </div>
        <div
          className="temperature"
        >
          -22
           ℃
        </div>
        <img
          alt="Weather icon"
          src="https://openweathermap.org/img/wn/icon_code.png"
        />
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Wind
        </div>
        <div
          className="value"
        >
          15 m/s
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Cloudness
        </div>
        <div
          className="value"
        >
          good to stay at home
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Pressure
        </div>
        <div
          className="value"
        >
          234 hPa
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Humidity
        </div>
        <div
          className="value"
        >
          1%
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Coords
        </div>
        <div
          className="value"
        >
          74, 89
        </div>
      </div>
      <button
        className="button"
        onClick={[Function]}
      >
        delete
      </button>
    </div>
  </div>
</div>
`;

exports[`<Cities /> render should has empty cities 1`] = `
<div
  className="cities"
>
  <h1>
    Сities
  </h1>
  <form
    class="add-city"
    onSubmit={[Function]}
  >
    <input
      class="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      class="button"
      type="submit"
      value="Add to city"
    />
  </form>
  <div
    className="forecasts"
  />
</div>
`;

exports[`<Cities /> render should has error and cities with 1 element 1`] = `
<div
  className="cities"
>
  <h1>
    Сities
  </h1>
  <form
    class="add-city"
    onSubmit={[Function]}
  >
    <input
      class="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      class="button"
      type="submit"
      value="Add to city"
    />
  </form>
  <div
    className="error"
  >
    Error: 
    test error message
  </div>
  <div
    className="forecasts"
  >
    <div
      className="city"
    >
      <div
        className="header"
      >
        <div
          className="city-name"
        >
          New York
        </div>
        <div
          className="temperature"
        >
          -22
           ℃
        </div>
        <img
          alt="Weather icon"
          src="https://openweathermap.org/img/wn/icon_code.png"
        />
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Wind
        </div>
        <div
          className="value"
        >
          15 m/s
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Cloudness
        </div>
        <div
          className="value"
        >
          good to stay at home
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Pressure
        </div>
        <div
          className="value"
        >
          234 hPa
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Humidity
        </div>
        <div
          className="value"
        >
          1%
        </div>
      </div>
      <div
        className="param"
      >
        <div
          className="name"
        >
          Coords
        </div>
        <div
          className="value"
        >
          74, 89
        </div>
      </div>
      <button
        className="button"
        onClick={[Function]}
      >
        delete
      </button>
    </div>
  </div>
</div>
`;

exports[`<Cities /> render should has error and empty cities 1`] = `
<div
  className="cities"
>
  <h1>
    Сities
  </h1>
  <form
    class="add-city"
    onSubmit={[Function]}
  >
    <input
      class="input"
      name="cityName"
      placeholder="City name"
      required={true}
      type="text"
    />
    <input
      class="button"
      type="submit"
      value="Add to city"
    />
  </form>
  <div
    className="error"
  >
    Error: 
    test error message
  </div>
  <div
    className="forecasts"
  />
</div>
`;
